mypy/checkexpr.py

```diff
@@ -2441,18 +2445,29 @@ def visit_index_expr_helper(self, e: IndexExpr) -> Type:
             # It's actually a type application.
             return self.accept(e.analyzed)
         left_type = self.accept(e.base)
-        if isinstance(left_type, TupleType) and self.chk.in_checked_function():
+        return self.visit_index_with_type(left_type, e)
+
+    def visit_index_with_type(self, left_type: Type, e: IndexExpr,
+                              original_type: Optional[Type] = None) -> Type:
+        """Analyze type of an index expression for a given type of base expression.
+
+        The 'original_type' is used for error messages (currently used for union types).
+        """
+        index = e.index
+        if isinstance(left_type, UnionType):
+            return UnionType.make_simplified_union([self.visit_index_with_type(typ, e, left_type)
```

Comment:
Just to check, should we pass `original_type` here as well?

Reply:
Normally, there should be no nested unions because they are flattened in the constructor, but I agree that it would be more robust to pass it also here.